DOC Add Maintainer Guidelines to the Contributing section (#9823)
silverstripe/silverstripe-framework: docs/en/05_Contributing/16_Maintainer_Guidelines.md

---
title: Maintainer guidelines
summary: Cheat sheet for module maintainers
iconBrand: wpforms
---

# Maintainer Guidelines

This doc is a cheat sheet for maintainers. It explains what that means to be a module maintainer and contains links to other resources with extra information.

Module maintainers are people taking care of the repositories, CI, documentation, source code, conventions, community communications, release management etc.

One of the most important maintainer responsibilities is to collaborate with other maintainers. Another important role is to facilitate community contributions (such as issue reports and pull requests).

[note]
A lot of extra information is available in Silverstripe CMS documentation section [“Contributing to Silverstripe”](./).
All maintainers should be familiar with those docs as they explain many details about how we work.
[/note]

## What is Silverstripe Core

> The "core" of Silverstripe is a set of non-dev dependencies contained in
> [silverstripe/recipe-cms](https://github.com/silverstripe/recipe-cms)
> and [silverstripe/recipe-core](https://github.com/silverstripe/recipe-cms).
> It is limited to dependencies owned by the "silverstripe" Packagist vendor,
> but can include "dev" dependencies.

Reviewer: ```suggestion
Silverstripe CMS is powered by a system of components in the form of Composer packages. These packages will be categorised as either a _module_ or a _recipe_.

The "core" of Silverstripe refers to the module packages owned by the "silverstripe" Packagist vendor that fall under one of the following recipes:

* [silverstripe/recipe-core](https://github.com/silverstripe/recipe-cms)
* [silverstripe/recipe-cms](https://github.com/silverstripe/recipe-cms)
* [silverstripe/installer](https://github.com/silverstripe/silverstripe-installer)
```

I've pulled this information from https://docs.silverstripe.org/en/4/getting_started/recipes/ and thought it might be a bit more clear.

## What are Supported Modules

In addition to Silverstripe Core, there are many [Supported Modules](https://www.silverstripe.org/software/addons/supported-modules-definition/)
which have the backing of Silverstripe Ltd. While it's a good idea to apply the rules outlined in this document,
work on these modules is guided by the 
[Supported Modules Standard](https://www.silverstripe.org/software/addons/supported-modules-definition/).
Commit access in Supported Modules is handled by agreement of the repository maintainers,
or any additional guidelines outlined via `README` and `CONTRIBUTING` files.

## Maintainer Roles

### Core Committers

The people looking after the Silverstripe Core modules.
See the details on the [Core Committers](./core_committers) page.

> #### House rules of the Core Committer team

Author: Moving to under `## Guidelines`.

There are mentions of `House Rules` here and there in the docs. If we rename the section to `Overview`, those references may lose their clarity. I suggest we rename it to `### House rules overview`.

### Contributing Committers

Beyond the Core Committer role, there can be individuals which
focus on core development work - typically sponsored through full-time product development roles by SilverStripe Ltd.
These Contributing Committers require write access to core repositories to maintain their pace,
often working alongside Core Committers. They are guided by additional rules:

 * Contributing Committers have write access to core repositories in order to work effectively with Github issues. They are expected to use those permissions with good judgement regarding merges of pull requests.
 * Complex or impactful changes need to be reviewed and approved by one or more Core Committers. This includes any additions, removals or changes to commonly used APIs. If that's not possible in the team, ping `@silverstripe/core-team` to get other Core Committers involved.
 * For these complex or impactful changes, Core Committers should be given 1-2 working days to review. Ideally at this point, the API has already been agreed on through issue comments outlining the planned work (see [RFC Process](request_for_comment]).

>  * More straightforward changes (e.g. documentation, styling) or areas which require quite specialised expertise (e.g. React) that's less available through most Core Committers can be approved or merged by team members who aren't Core Committers

Reviewer: ```suggestion
 * More straightforward changes (e.g. documentation, styling) or areas which require quite specialised expertise (e.g. React) that's less available through most Core Committers can be approved or merged by team members who aren't Core Committers.
```

[warning]
Some of the Supported Modules have their own house rules (usually explained in the repository README or CONTRIBUTING files). Please remember to read those carefully.
[/warning]

### How to triage

Triaging GitHub issues and pull requests imply responding to issue/pr creators with following closing or labelling. Labelling allows the other maintainers to pick up proper issues/prs accordingly to their dedication or available resources.

It is also the main way to escalate critical and high impact issues, give feedback to the community, or peer review a PR and let other maintainers know it is “good to be merged on green CI tests”.

How to do it
  - Regularly check the repository for new issues
  - Ideally subscribe to the repository via [GitHub Watch](https://docs.github.com/en/free-pro-team@latest/github/managing-subscriptions-and-notifications-on-github/viewing-your-subscriptions#configuring-your-watch-settings-for-an-individual-repository) functionality
  - Read the docs about [how we use labels](./code/#labels)
  - If unsure about anything, it’s usually good to ask other maintainers for their opinions (on Slack or via mentioning them directly on GitHub)

> 

Reviewer: Is it worth adding reference to the 'Untriaged' tab at https://silverstripe-github-issues.now.sh/?# ?

Reviewer 2: Done above

### Triage

Triage of issues and pull request is an important activity: Reviewing issues, adding labels,
closing stale issues, etc. This does not require write access to the repository as a "Contributing Committer".
This is a great way for active community members to help out, and start a path towards becoming a Core Committer.

Triage roles may be granted by any Core Committer,
who should give other Core Committers an opportunity to weigh in on the decision.

## The Guidelines

With great power (write access) comes great responsibility.

First and foremost rule of a maintainer is to collaborate with other maintainers. Follow the house rules and remember that others also care about the modules they maintain.

Remember to follow the [House rules of the Core Committer team](#house-rules-of-the-core-committer-team) and the supported module [guidelines](https://www.silverstripe.org/software/addons/supported-modules-definition/).

[warning]
Some of the Supported Modules have their own house rules (usually explained in the repository README or CONTRIBUTING files). Please remember to read those carefully.
[/warning]

### How to triage

> Triaging GitHub issues and pull requests imply responding to issue/pr creators with following closing or labelling. Labelling allows the other maintainers to pick up proper issues/prs accordingly to their dedication or available resources.

Reviewer 2: ```suggestion
Triaging issues and pull requests involves review, applying labels, as well as closing invalid submissions.
```

> It is also the main way to escalate critical and high impact issues, give feedback to the community, or peer review a PR and let other maintainers know it is “good to be merged on green CI tests”.

Reviewer 2: ```suggestion
Applying [labels](./code/#labels) is the main mechanism used to escalate critical and high impact issues, give feedback to the community, or peer review a PR and let other maintainers know it is “good to be merged on green CI tests”.
```

How to do it

>   - Regularly check the repository for new issues

Reviewer 2: ```suggestion
  - Regularly check the repository for new issues or use the ["Untriaged" filter](https://silverstripe-github-issues.now.sh/?mode=untriaged) in our cross-repository issue tracker.
```
